docs/library/network.WLAN.rst

```diff
@@ -15,13 +15,28 @@ This class provides a driver for WiFi network processors.  Example usage::
 
 Constructors
 ------------
-.. class:: WLAN(interface_id)
+.. class:: WLAN(interface_id [, wiring=parameter_set])
 
 Create a WLAN network interface object. Supported interfaces are
 ``network.STA_IF`` (station aka client, connects to upstream WiFi access
 points) and ``network.AP_IF`` (access point, allows other WiFi clients to
 connect). Availability of the methods below depends on interface type.
 For example, only STA interface may `WLAN.connect()` to an access point.
+At ports which allow to connect an external WLAN device like the
+NINA W102, the optional parameter ``wiring`` can be supplied
+to define the physical connection to the WLAN module, like the SPI
+object and control pins. The number and type of parameters depend
+on the WLAN module and are checked by it's driver. Typically 
+``wiring`` is a tuple of several items. Example for the NINA W102::
+
+    import network
+    # enable station interface with wiring
+    # The value for the GP0 pin can be set to `None`.
+    nic = network.WLAN(network.STA_IF, wiring=(SPI_obj, CS_pin, Busy_pin, RST_pin, GP0_pin))
+    # or if the GP0 pin is not wired, since it is not needed for WLAN
+    nic = network.WLAN(network.STA_IF, wiring=(SPI_obj, CS_pin, Busy_pin, RST_pin, None))
```

Comment:
I discussed this with @jimmo.  To solve this problem properly will take a fair bit of effort (in API design) because it needs to support the general case of (for example) multiple WLAN drivers of different types, eg a board with both an external Nina-based module and cyw43-based module.  How would one specify these different drivers?  Maybe `network.NinaW10` and `network.CYW43`?  And how would they work together since the former doesn't use lwIP while the latter does?

All we really need to support here is the Adafruit Airlift, which can be done simply by a compile-time configuration and available as `network.WLAN()`.  Ie it doesn't need any arguments to configure it, it just uses the default settings which are fixed for the Airlift board.

If someone wants to use a Nina-based module (esp32) on some other samd-based board then it's pretty simple to make your own board definition and build it.  Then `network.WLAN()` will just work without any arguments because it's all configured at the C level.

In the future we can support the more general cases of dynamically (at runtime) instantiating WLAN interfaces of various types, but for now, to keep things moving and without getting bogged down in (long) API discussions, I suggest to just keep it simple and only allow the settings (SPI bus etc) be configured at compile time.

Reply:
> All we really need to support here is the Adafruit Airlift, which can be done simply by a compile-time configuration and available as network.WLAN(). Ie it doesn't need any arguments to configure it, it just uses the default settings which are fixed for the Airlift board.

It's not only Adafruit airlift, it's the general case. And even with Adafruit airlift there are various alternatives. There are:
- Boards with the airlift built-on, like the Metro M4 airlift. There, the wiring is fixed.
- Dedicated Add-on modules for the Itsy-Bitsy and Feather device, where the wiring is fixed as well.
- There is a UNO-Shaped Add-on board with fixed wiring, which works as well on some MIMXRT board, with a suggested wiring, which can be changed. I wired one up using a genuine ESP32 to be tested with MIMXRT boards. Works well at boards without Ethernet. If there is Ethernet, there is a conflict with the LWIP stack.
- And there are simple Adafruit and ESP32 breakout boards. I use one with a Sparkfun SAMD board. For these you could make a certain wiring mandatory. 
- SAMD boards for which no matching WiFi add.-on exists which still could use a simple ESP32 based breakout. For these boards a preferred wiring may not fit, just because the pins are not available or needed for a different purpose.

The way the code is implemented now, you can just write `network.WLAN()` if there is a wiring defined for the board and it is to be used. If one want to use a different type of connection, it must be specified as `network.WLAN(interface, spi, cs, busy, reset)` (GP0 is only required for Firmware update). So both options exist. And we are discussing here only about the NINAW10 network driver. So the question of having more than one WiFi is a different topic.

So I would suggest to (re-)add a preferred wiring to the SAMD board collection and keep the option of specifying an alternative wiring.

Reply:
@dpgeorge This update of the PR covers the default wiring schemes for the boards, where is is applicable, and I removed the changes to BLE() in the documentation. The Metro M4 Airlift PR had already the required defines, so it is not added here. The wiring for the Grand Central M4 Epress using the Airlift shield can be added as soon as the board files for this board is merged PR #11104.

Reply:
> The way the code is implemented now, you can just write `network.WLAN()` if there is a wiring defined for the board and it is to be used. If one want to use a different type of connection, it must be specified as `network.WLAN(interface, spi, cs, busy, reset)`

I think this is close but needs a small change:
- If there is a default driver and wiring (eg NINA with specific pins like on the Airlift), then indeed `network.WLAN()` can be used
- Otherwise, one must be explicit with both the driver and the wiring.  I don't think it's a good idea to document that `network.WLAN()` can take a large set of keyword arguments, for all the different possible WLAN drivers.  Instead it would be more future proof to instantiate the driver from a class related to that driver, eg `network.NinaW10(...)`.

The documentation for `network.WLAN()` should not change.  Documentation for `network.NinaW10()` and its supported arguments (spi etc) would be added.

For boards like PYBD and Pico W the underlying WLAN driver would be `network.CYW43()` (although that class doesn't have to exist in the `network` module, `WLAN` can just be an alias to it, which is actually how it works at the moment).

If a samd board supports both a default WLAN and custom NINAW10 wiring, then it can have both `network.WLAN()` and `network.NinaW10()`.

------

Also, I don't think we should enable the NINAW10 driver by default on all the SAMD51 boards.  It adds a lot of code, and sets a precedent that's hard to back down from later on.  If you really want to provide pre-built firmware that has NINWA10 enabled then we can instead try adding board variants (like on stm32), so there is a base variant of the samd boards without networking, and a variant with networking included.

Reply:
> Documentation for network.NinaW10() and its supported arguments (spi etc) would be added.

Do you suggest adding network.NinaW10() as alias to network.WLAN(), so both can be used and the difference is only in the documentation?
___________________________
> Also, I don't think we should enable the NINAW10 driver by default on all the SAMD51 boards. It adds a lot of code, and sets a precedent that's hard to back down from later on. 

I agree that not all SAMD51 board can handle that, mechanically and by flash size. The size of the code is: Core WLAN drivers ~15k, MBEDTLS ~100k, network utility package ~10k. So the basic WLAN driver w/o encryption is small. And the code size may be less of a problem if support for external flash is present (PR #10233). But having variants is a good suggestion.

Reply:
> Do you suggest adding network.NinaW10() as alias to network.WLAN(), so both can be used and the difference is only in the documentation?

In effect, yes that would be the change.

The idea is:
- If a board has built-in wlan then everyone just uses `network.WLAN()` to get at the interface and they don't need to think any further.
- If a board doesn't have built-in wlan (or lan, ble, etc) then one needs to think which driver they need and what connections to use.  The particular driver is specified by the class name.

Reply:
Fine. I'll set that up.